tensorflow_model_optimization/python/core/sparsity_tf2/sparse_utils_test.py

```diff
@@ -0,0 +1,164 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Tests for the utilities in sparse training pipelines."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from absl.testing import parameterized
+import numpy as np
+import tensorflow as tf
+
+from tensorflow.python.keras import keras_parameterized
+from tensorflow_model_optimization.python.core.sparsity_tf2 import sparse_utils
+
+dtypes = tf.dtypes
+test = tf.test
+
+
+def sample_noise(shape=(), mu=0, sigma=1.):
+  sample = tf.random.normal(shape, mean=mu,  stddev=sigma, dtype=tf.float64)
+  return sample
+
+def calculate_stdev(p):
+  return np.sqrt(p * (1 - p))
+
+class SparseUtilsTest(test.TestCase, parameterized.TestCase):
+
+  def setUp(self):
+    super(SparseUtilsTest, self).setUp()
+    self.ratios = [0.0, 0.1, 0.25, 0.5, 0.75, 0.9, 1.0]
+    self.shapes = [(), [1], [5], [2, 2], [4, 4], [8, 4], [3, 17], [3, 6, 9]]
+    self.seeds = range(20)
+
+  def matrix_init(self, _shape):
+    return tf.ones(_shape)
+  
+  def matrix_init_with_type(self, _shape, _type):
+    tf.ones(_shape, dtype=_type)
+
+  def get_bernouilli(self, p):
+    return sparse_utils.Bernouilli(p)
+
+  def get_permuteones(self, ratio):
+    return sparse_utils.PermuteOnes(ratio)
+
+  @parameters.parameters(
+    (-1.,), (1.5,)
+  )
+  def testInvalidRatioRaisesError(self, p):
+    with self.assertRaises(ValueError):
+      bernouilli = self.get_bernouilli(p)
+      permuteones = self.get_permuteones(p)
+
+  @parameterized.parameters(
+    (i,) for i in self.ratios
+  )
+  def testScatteredInitialOnes(self, ratio):
+    shape = (4, 4)
+    matrix = self.matrix_init(shape)
+    self.assertAllEqual(shape, matrix.shape)
+
+    def _check_scatter(sparse_matrix):
+      ones_indices = np.where(sparse_matrix == 1.)
+      mean_nonzeros = np.mean(ones_indices)
+      ones_count = np.count_nonzero(sparse_matrix)
+      return ones_count, mean_nonzeros
+
+    seed = 0
+    ones_indices = []
+    ones_counts = []
+    for _ in range(100):
+      matrix_sparse_permute = self.get_permuteones(ratio)(shape, seed=seed)
+      ones_count, mean_nonzeros = _check_scatter(matrix_sparse_permute)
+      ones_indices.append(mean_nonzeros)
+      ones_counts.append(ones_count)
+      seed += 1
+    # check that 1/0s are scattered
+    # mean of entire matrix and mean of entire nonzeros
+    self.assertAllClose(ones_indices[:len(ones_indices)//2], ones_indices[len(ones_indices)//2:])
```

Comment:
This check is really confusing to interpet, and seems pretty wrong?

Isn't this checking that the first half of the trials match the second half of the trials?

Just check that the mean matrix (if you reduce the initialized values across all trials but leave it as a matrix) approximately matches a tensor of the corresponding input shape where all the values are `ratio`.